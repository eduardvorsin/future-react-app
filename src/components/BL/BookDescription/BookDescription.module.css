.book-description {
	display: grid;
	grid-template-rows: min-content;
	grid-auto-rows: min-content;
	align-items: start;
	gap: 30px;
}

.book-description__main-title {
	line-height: 1.25;
	grid-column: -1/1;
	transition:
		color var(--transition-time) var(--transition-easing);
}

.book-description__img-box {
	max-width: 300px;
	height: auto;
	width: 100%;
	justify-self: center;
}

.book-description__img {
	background-color: var(--brand-color-100);
	border: 2px solid var(--brand-color-100);
	width: 100%;
	height: 100%;
	object-fit: cover;
}

.book-description__content {
	/* styles for .book-description__content */
}

.book-description__title {
	text-align: left;
	margin-bottom: 10px;
	transition:
		color var(--transition-time) var(--transition-easing);
}

.book-description__category {
	margin: 0;
	font-size: 18px;
	margin-bottom: 20px;
	color: var(--text-color-primary);
	transition:
		color var(--transition-time) var(--transition-easing);
}

.book-description__category>span {
	font-weight: 700;
}

.book-description__preview-link {
	margin-bottom: 15px;
}

.book-description__authors {
	margin: 0;
	margin-bottom: 25px;
}

.book-description__text {
	color: var(--text-color-primary);
	margin: 0;
	transition:
		color var(--transition-time) var(--transition-easing);
}

.book-description__buying {
	display: flex;
	flex-direction: column;
	align-items: flex-start;
	border-radius: 8px;
	box-shadow: hsla(240deg, 5%, 41%, 20%) 0 7px 29px 0;
	padding: 15px;
	max-width: 250px;
	transition:
		box-shadow var(--transition-time) var(--transition-easing),
		background-color var(--transition-time) var(--transition-easing);
}

.book-description__price {
	margin: 0;
	font-size: 26px;
	margin-bottom: 10px;
	color: var(--text-color-primary);
	transition:
		color var(--transition-time) var(--transition-easing);
}

.book-description__buying-btn {
	/* styles for .book-description__buying-btn */
}

.book-description__describing {
	grid-column: 1/-1;
}

@media (min-width:576px) {
	.book-description__img-box {
		max-width: 400px;
	}

	.book-description {
		grid-template-columns: 1fr;
	}

	.book-description__title {
		margin-bottom: 15px;
	}
}

@media (min-width:768px) {
	.book-description {
		grid-template-columns: 1fr 1.5fr;
	}

	.book-description__buying {
		grid-column: 1/2;
	}

	.book-description__describing {
		grid-column: 2/3;
	}

	.book-description__img-box {
		max-width: none;
		justify-self: auto;
	}
}

@media (min-width:992px) {
	.book-description {
		grid-template-columns: 1fr 2fr;
	}
}

@media(min-width:1200px) {
	.book-description {
		grid-template-columns: 1.5fr 1.5fr 250px;
		gap: 30px 20px;
	}

	.book-description__describing {
		grid-column: 1/-1;
	}

	.book-description__img-box {
		width: auto;
		margin-right: 20px;
	}

	.book-description__buying {
		grid-column: 3/4;
	}
}

@media(prefers-color-scheme:dark) {
	.book-description__buying {
		box-shadow: none;
		background-color: var(--secondary-color-200);
		transition:
			box-shadow var(--transition-time) var(--transition-easing),
			background-color var(--transition-time) var(--transition-easing);
	}
}