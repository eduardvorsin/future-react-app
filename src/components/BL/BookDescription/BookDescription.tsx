import React, { FC } from 'react';
import { useLoaderData } from 'react-router-dom';
import { useTranslation } from 'react-i18next';
import Title from '../../UI/Title/Title';
import classes from './BookDescription.module.css';
import PlaceholderImageRu from '../../../assets/images/placeholder388x613-ru.jpeg';
import PlaceholderImageEn from '../../../assets/images/placeholder388x613-en.jpeg';
import Link from '../../UI/Link/Link';
import OpenBookIcon from '../../../assets/images/icons/open-book.svg';
import currencyFormat from '../../../localization/formatting/numbers';
import CharacteristicList from '../../UI/CharacteristicList/CharacteristicList';
import bookDescriptionLoader from '../../../router/loaders/bookDescriptionLoader';

type BookDescriptionProps = {
  testId?: string,
};

const BookDescription: FC<BookDescriptionProps> = ({
  testId,
}) => {
  const { t, i18n } = useTranslation();
  const localizedPlaceholder = i18n.language === 'ru' ? PlaceholderImageRu : PlaceholderImageEn;
  const currentBook = useLoaderData() as Awaited<ReturnType<typeof bookDescriptionLoader>>;

  const categories = currentBook?.categories?.[0] ?? t('unknown');
  const imageSrc = currentBook?.imageLinks?.medium ?? localizedPlaceholder;
  const description = currentBook?.description ?? t('notAvaliableDescription', { ns: 'bookDescription' });
  const title = currentBook?.title ?? t('unknown');
  const previewLink = currentBook?.webReaderLink ?? '';

  let price = currentBook?.retailPrice?.amount ?? t('notAvailable');

  if (typeof price === 'number') {
    price = currencyFormat(price, {
      currency: currentBook?.retailPrice?.currencyCode,
    });
  }
  return (
    <div
      className={classes['book-description']}
      data-testid={testId}
    >
      <Title
        className={classes['book-description__main-title']}
        component='h2'
        variant='dark'
        level={3}
      >
        {title}
      </Title>
      <div
        className={classes['book-description__img-box']}
      >
        <img
          src={imageSrc}
          className={classes['book-description__img']}
          alt={`${title} ${t('bookCover')}`}
          width={435}
          height={632}
        />
      </div>
      <div
        className={classes['book-description__content']}
      >
        <Link
          className={classes['book-description__preview-link']}
          icon={OpenBookIcon}
          path={previewLink}
        >
          {t('readBookFragment', { ns: 'bookDescription' })}
        </Link>
        <p
          className={classes['book-description__category']}
        >
          {t('categories', { ns: 'bookDescription' })}:
          <span>{categories}</span>
        </p>
        {currentBook && (
          <>
            <Title
              className={classes['book-description__title']}
              component='h3'
              variant='dark'
              level={4}
            >
              {t('characteristicsTitle', { ns: 'bookDescription' })}
            </Title>

            <CharacteristicList
              bookData={currentBook}
            />
          </>
        )}
      </div>
      <div className={classes['book-description__buying']}>
        <p
          className={classes['book-description__price']}
        >
          {price}
        </p>
        <Link
          path={currentBook?.buyLink}
          variant='tertiary'
          className={classes['book-description__buying-btn']}
        >
          {t('buy')}
        </Link>
      </div>
      <div
        className={classes['book-description__describing']}
      >
        <Title
          className={classes['book-description__title']}
          component='h3'
          variant='dark'
          level={4}
        >
          {t('descriptionTitle', { ns: 'bookDescription' })}
        </Title>
        <p
          className={classes['book-description__text']}
        >
          {description}
        </p>
      </div>

    </div>
  );
};

export default BookDescription;
