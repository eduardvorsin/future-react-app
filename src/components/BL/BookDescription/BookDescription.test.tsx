import React from 'react';
import { render, waitFor, screen } from '@testing-library/react';
import { RouterProvider, createMemoryRouter } from 'react-router-dom';
import BookDescription from './BookDescription';
import ReduxWrapper from '../../../tests/helpers/ReduxWrapper';
import i18n from '../../../localization/i18next';

const bookDataMock = {
  authors: [
    'Henchett Eric',
    'Listuon Benjamin',
  ],
  categories: ['Computers / Languages / JavaScript'],
  description: 'Vue.js - this is a popular library for creating user interfaces.',
  language: 'en',
  maturityRating: 'NOT_MATURE',
  pageCount: 304,
  publishedDate: '2019-02-25',
  publisher: '"Издательский дом ""Питер"""',
  title: 'Vue.js в действии',
  industryIdentifiers: [
    {
      type: 'ISBN_10',
      identifier: '5446110986',
    },
    {
      type: 'ISBN_13',
      identifier: '9785446110988',
    },
  ],
  imageLinks: {
    smallThumbnail: 'http://books.google.com/books/publisher/content?id=ZGGLDwAAQBAJ&printsec=frontcover&img=1&zoom=5&edge=curl&imgtk=AFLRE71e3hoeDHHQQLo88dJ5CA-_YYDITEiOpKKp6ar4opPxAckmKLwAl6TQxsjPQH-F9zVgoStRHL1S8NpPD1N7yGe5BFdMQSzrgNTVmJYnNdfrSr9yGW2agtlbQ5NHgYMxj--9qJe7&source=gbs_api',
    thumbnail: 'http://books.google.com/books/publisher/content?id=ZGGLDwAAQBAJ&printsec=frontcover&img=1&zoom=1&edge=curl&imgtk=AFLRE71Dt516axDZ-SnR2stjeuBpPtptWBbMDg16UMmAqiSPOphOTWqQIfDdRxUYQKymbU8wPk4XJMgFZTfZgmqhxM073I6IIkYDr-Ewk46ITSYfPGCxSsEW9Tz_fupqZHEvs5bgH2NN&source=gbs_api',
    small: 'http://books.google.com/books/publisher/content?id=ZGGLDwAAQBAJ&printsec=frontcover&img=1&zoom=2&edge=curl&imgtk=AFLRE704N6ZtYHbUfTMclBu28YCI60N-jQpNpzAy4ORmlidrrfKJqKAZF6fHZwdUzLfesFycT64GvFUgTrIQkM8JgXG60UBcy0oGIUlvBHib-wggpfIOOQEsHrzUw19ZipV8EVRTJEsh&source=gbs_api',
    medium: 'http://books.google.com/books/publisher/content?id=ZGGLDwAAQBAJ&printsec=frontcover&img=1&zoom=3&edge=curl&imgtk=AFLRE73kEchC1gD2nAmiV_Gnyc8Fxl5nuJayRpTus8CUWPeH5E-YU9NEZmUciVSF8VpxZvuRlcRwcBkenVrC1O77GBB9QaGCM_3AB3pDNwjRDzJMVm-R77-XcoHeVQ557ksVjkWfatkA&source=gbs_api',
    large: 'http://books.google.com/books/publisher/content?id=ZGGLDwAAQBAJ&printsec=frontcover&img=1&zoom=4&edge=curl&imgtk=AFLRE70NnA4BuexkY5ND8jA54P_siFmRilL6IlKsLP-vduWOLxB8vrp3d2wSDBr170TSZFLdb_uQOUv19ngcSGeiqwqdstTAT6HknjLk-X77CA-JXIm1SgJtB_VwKz3KMR878VPeYMpJ&source=gbs_api',
    extraLarge: 'http://books.google.com/books/publisher/content?id=ZGGLDwAAQBAJ&printsec=frontcover&img=1&zoom=6&edge=curl&imgtk=AFLRE70JbpYFxiRvR1scLtR4JvU3L6_xM1SWLBgL0HyGqYQX0ZADI9WIO8XEliJysMjF4WpbxaTN358BPAQQJzWhcXKpNWLuLKBIsra9hPA7Ii9VaLX1rrvGvCGDqPHMmvDvmCa5i1yr&source=gbs_api',
  },
  webReaderLink: 'http://play.google.com/books/reader?id=ZGGLDwAAQBAJ&hl=&source=gbs_api',
  buyLink: 'https://play.google.com/store/books/details?id=ZGGLDwAAQBAJ&rdid=book-ZGGLDwAAQBAJ&rdot=1&source=gbs_api',
  isEbook: true,
  retailPrice: {
    amount: 6593.15,
    currencyCode: 'KZT',
  },
};

const routes = [{
  path: '/books/:id',
  loader: async () => bookDataMock,
  element: (
    <BookDescription
      testId='test-book-description'
    />
  ),
}];
const router = createMemoryRouter(routes, { initialEntries: ['/books/ZGGLDwAAQBAJ'] });

describe('BookDescription test', () => {
  it('should render correctly', async () => {
    render(<RouterProvider router={router} />, { wrapper: ReduxWrapper });

    await waitFor(() => {
      expect(screen.getByTestId<HTMLDivElement>('test-book-description')).toBeInTheDocument();
    });
  });

  it('if an incorrect id is specified in the url, the component will render placeholders for book data', async () => {
    const routerWithEmptyData = createMemoryRouter([{
      path: '/books/:id',
      loader: async () => null,
      element: (
        <BookDescription
          testId='test-book-description'
        />
      ),
    }], { initialEntries: ['/books/ZGGLDwAAQBAJ'] });
    render(<RouterProvider router={routerWithEmptyData} />, { wrapper: ReduxWrapper });

    await waitFor(() => {
      expect(screen.getAllByText<HTMLParagraphElement>(/unknown/i)).toHaveLength(2);
    });

    expect(screen.getAllByText<HTMLParagraphElement>(/not available/i)).toHaveLength(2);
  });

  it('is a basic snapshot', async () => {
    render(<RouterProvider router={router} />, { wrapper: ReduxWrapper });

    await waitFor(() => {
      expect(screen.getByTestId<HTMLDivElement>('test-book-description')).toBeInTheDocument();
    });

    expect(screen.getByTestId<HTMLDivElement>('test-book-description')).toMatchSnapshot();
  });

  it('is a snapshot with placeholders for missing data', async () => {
    const routerWithEmptyData = createMemoryRouter([{
      path: '/books/:id',
      loader: async () => null,
      element: (
        <BookDescription
          testId='test-book-description'
        />
      ),
    }], { initialEntries: ['/books/ZGGLDwAAQBAJ'] });
    render(<RouterProvider router={routerWithEmptyData} />, { wrapper: ReduxWrapper });

    await waitFor(() => {
      expect(screen.getAllByText<HTMLParagraphElement>(/unknown/i)).toHaveLength(2);
    });

    expect(screen.getByTestId<HTMLDivElement>('test-book-description')).toMatchSnapshot();
  });
});

describe('BookDescription integration tests', () => {
  it('the component with the text in English is displayed correctly', () => {
    const bookFragmentLink = new RegExp(i18n.t('readBookFragment', { ns: 'bookDescription' }));
    const categories = new RegExp(i18n.t('categories', { ns: 'bookDescription' }));
    const characteristicsTitle = new RegExp(i18n.t('characteristicsTitle', { ns: 'bookDescription' }));
    const buyLink = new RegExp(i18n.t('buy'));

    render(<RouterProvider router={router} />, { wrapper: ReduxWrapper });

    expect(screen.getByText<HTMLAnchorElement>(bookFragmentLink)).toBeInTheDocument();
    expect(screen.getByText<HTMLParagraphElement>(categories)).toBeInTheDocument();
    expect(screen.getByText<HTMLHeadingElement>(characteristicsTitle)).toBeInTheDocument();
    expect(screen.getByText<HTMLAnchorElement>(buyLink)).toBeInTheDocument();
  });

  it('the component with the text in Russian is displayed correctly', () => {
    i18n.changeLanguage('ru');
    const bookFragmentLink = new RegExp(i18n.t('readBookFragment', { ns: 'bookDescription' }));
    const categories = new RegExp(i18n.t('categories', { ns: 'bookDescription' }));
    const characteristicsTitle = new RegExp(i18n.t('characteristicsTitle', { ns: 'bookDescription' }));
    const buyLink = new RegExp(i18n.t('buy'));

    render(<RouterProvider router={router} />, { wrapper: ReduxWrapper });

    expect(screen.getByText<HTMLAnchorElement>(bookFragmentLink)).toBeInTheDocument();
    expect(screen.getByText<HTMLParagraphElement>(categories)).toBeInTheDocument();
    expect(screen.getByText<HTMLHeadingElement>(characteristicsTitle)).toBeInTheDocument();
    expect(screen.getByText<HTMLAnchorElement>(buyLink)).toBeInTheDocument();
  });
});
