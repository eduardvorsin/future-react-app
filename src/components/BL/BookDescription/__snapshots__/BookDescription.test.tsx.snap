// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookDescription test is a basic snapshot 1`] = `
<div
  class="book-description"
  data-testid="test-book-description"
>
  <h2
    class="title book-description__main-title title--dark title--level-3"
  >
    Vue.js в действии
  </h2>
  <div
    class="book-description__img-box"
  >
    <img
      alt="Vue.js в действии book cover"
      class="book-description__img"
      height="632"
      src="http://books.google.com/books/publisher/content?id=ZGGLDwAAQBAJ&printsec=frontcover&img=1&zoom=3&edge=curl&imgtk=AFLRE73kEchC1gD2nAmiV_Gnyc8Fxl5nuJayRpTus8CUWPeH5E-YU9NEZmUciVSF8VpxZvuRlcRwcBkenVrC1O77GBB9QaGCM_3AB3pDNwjRDzJMVm-R77-XcoHeVQ557ksVjkWfatkA&source=gbs_api"
      width="435"
    />
  </div>
  <div
    class="book-description__content"
  >
    <a
      aria-disabled="false"
      class="link book-description__preview-link  link--primary"
      href="http://play.google.com/books/reader?id=ZGGLDwAAQBAJ&hl=&source=gbs_api"
      rel="noopener"
    >
      <span
        aria-hidden="true"
        class="link__icon-wrapper link__icon-wrapper--left"
        data-testid="icon-wrapper"
      >
        <svg />
      </span>
      Read a fragment of the book
    </a>
    <p
      class="book-description__category"
    >
      Categories
      :
      <span>
        Computers / Languages / JavaScript
      </span>
    </p>
    <h3
      class="title book-description__title title--dark title--level-4"
    >
      Characteristics
    </h3>
    <ul
      class="characteristic-list "
    >
      <li
        class="characteristic-list__item"
      >
        <span
          class="characteristic-list__property"
        >
          Authors
        </span>
        <span
          class="characteristic-list__value"
        >
          Henchett Eric, Listuon Benjamin
        </span>
      </li>
      <li
        class="characteristic-list__item"
      >
        <span
          class="characteristic-list__property"
        >
          Language
        </span>
        <span
          class="characteristic-list__value"
        >
          English
        </span>
      </li>
      <li
        class="characteristic-list__item"
      >
        <span
          class="characteristic-list__property"
        >
          Content 18+
        </span>
        <span
          class="characteristic-list__value"
        >
          Yes
        </span>
      </li>
      <li
        class="characteristic-list__item"
      >
        <span
          class="characteristic-list__property"
        >
          Page count
        </span>
        <span
          class="characteristic-list__value"
        >
          304
        </span>
      </li>
      <li
        class="characteristic-list__item"
      >
        <span
          class="characteristic-list__property"
        >
          Published date
        </span>
        <span
          class="characteristic-list__value"
        >
          2019-02-25
        </span>
      </li>
      <li
        class="characteristic-list__item"
      >
        <span
          class="characteristic-list__property"
        >
          Publisher
        </span>
        <span
          class="characteristic-list__value"
        >
          "Издательский дом ""Питер"""
        </span>
      </li>
      <li
        class="characteristic-list__item"
      >
        <span
          class="characteristic-list__property"
        >
          ISBN
        </span>
        <span
          class="characteristic-list__value"
        >
          9785446110988
        </span>
      </li>
      <li
        class="characteristic-list__item"
      >
        <span
          class="characteristic-list__property"
        >
          Book Type
        </span>
        <span
          class="characteristic-list__value"
        >
          E-book
        </span>
      </li>
    </ul>
  </div>
  <div
    class="book-description__buying"
  >
    <p
      class="book-description__price"
    >
      KZT 6,593
    </p>
    <a
      aria-disabled="false"
      class="link book-description__buying-btn  link--tertiary"
      href="https://play.google.com/store/books/details?id=ZGGLDwAAQBAJ&rdid=book-ZGGLDwAAQBAJ&rdot=1&source=gbs_api"
      rel="noopener"
    >
      Buy
    </a>
  </div>
  <div
    class="book-description__describing"
  >
    <h3
      class="title book-description__title title--dark title--level-4"
    >
      Description
    </h3>
    <p
      class="book-description__text"
    >
      Vue.js - this is a popular library for creating user interfaces.
    </p>
  </div>
</div>
`;

exports[`BookDescription test is a snapshot with placeholders for missing data 1`] = `
<div
  class="book-description"
  data-testid="test-book-description"
>
  <h2
    class="title book-description__main-title title--dark title--level-3"
  >
    Unknown
  </h2>
  <div
    class="book-description__img-box"
  >
    <img
      alt="Unknown book cover"
      class="book-description__img"
      height="632"
      src="test-file-stub"
      width="435"
    />
  </div>
  <div
    class="book-description__content"
  >
    <a
      aria-disabled="true"
      class="link book-description__preview-link link--disabled link--primary"
      href=""
      rel="noopener"
      role="link"
    >
      <span
        aria-hidden="true"
        class="link__icon-wrapper link__icon-wrapper--left"
        data-testid="icon-wrapper"
      >
        <svg />
      </span>
      Read a fragment of the book
    </a>
    <p
      class="book-description__category"
    >
      Categories
      :
      <span>
        Unknown
      </span>
    </p>
  </div>
  <div
    class="book-description__buying"
  >
    <p
      class="book-description__price"
    >
      Not available
    </p>
    <a
      aria-disabled="true"
      class="link book-description__buying-btn link--disabled link--tertiary"
      rel="noopener"
      role="link"
    >
      Buy
    </a>
  </div>
  <div
    class="book-description__describing"
  >
    <h3
      class="title book-description__title title--dark title--level-4"
    >
      Description
    </h3>
    <p
      class="book-description__text"
    >
      The description is not available for this book
    </p>
  </div>
</div>
`;
